File: src/aeonik_ingrain/security.py
"""Small safety helpers for local learned-experience storage."""

from __future__ import annotations

import re

SECRET_PATTERNS = [
    re.compile(r"(?i)(api[_-]?key|secret|token|password)\s*[:=]\s*['\"]?([A-Za-z0-9_\-\.]{16,})"),
    re.compile(r"sk-[A-Za-z0-9]{20,}"),
    re.compile(r"ghp_[A-Za-z0-9]{20,}"),
    re.compile(r"xox[baprs]-[A-Za-z0-9\-]{20,}"),
    re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----"),
]

INJECTION_MARKERS = (
    "ignore previous instructions",
    "disregard previous instructions",
    "system prompt",
    "developer message",
    "you are now",
    "act as system",
)

INVISIBLE_RE = re.compile(r"[\u200b\u200c\u200d\ufeff]")


def strip_invisible(text: str) -> str:
    return INVISIBLE_RE.sub("", text or "")


def _redact_match(match: re.Match[str]) -> str:
    if match.lastindex and match.lastindex >= 1:
        key = match.group(1)
        return f"{key}=<redacted>"
    return "<redacted-secret>"


def redact_secrets(text: str) -> str:
    value = strip_invisible(text)
    for pattern in SECRET_PATTERNS:
        value = pattern.sub(_redact_match, value)
    return value


def has_likely_secret(text: str) -> bool:
    value = strip_invisible(text or "")
    return any(pattern.search(value) for pattern in SECRET_PATTERNS)


def has_prompt_injection_marker(text: str) -> bool:
    lower = (text or "").lower()
    return any(marker in lower for marker in INJECTION_MARKERS)


def sanitize_for_storage(text: str, *, limit: int = 20000) -> str:
    value = redact_secrets(text or "").strip()
    if len(value) > limit:
        value = value[:limit].rstrip() + "\n[... truncated by Ingrain]"
    return value


def sanitize_for_context(text: str) -> str:
    value = redact_secrets(strip_invisible(text or "").strip())
    if has_prompt_injection_marker(value):
        return "[possible prompt-injection text withheld by Ingrain; inspect source event before trusting]"
    return value

File: crates/ingrain-core/src/security.rs
const INJECTION_MARKERS: [&str; 6] = [
    "ignore previous instructions",
    "disregard previous instructions",
    "system prompt",
    "developer message",
    "you are now",
    "act as system",
];

const WITHHELD: &str =
    "[possible prompt-injection text withheld by Ingrain; inspect source event before trusting]";

pub fn sanitize_for_context(text: &str) -> String {
    let stripped = strip_invisible(text).trim().to_string();
    let redacted = redact_obvious_secrets(&stripped);
    if has_prompt_injection_marker(&redacted) {
        WITHHELD.to_string()
    } else {
        redacted
    }
}

fn strip_invisible(text: &str) -> String {
    text.chars()
        .filter(|ch| !matches!(ch, '\u{200b}' | '\u{200c}' | '\u{200d}' | '\u{feff}'))
        .collect()
}

fn has_prompt_injection_marker(text: &str) -> bool {
    let lower = text.to_lowercase();
    INJECTION_MARKERS
        .iter()
        .any(|marker| lower.contains(marker))
}

fn redact_obvious_secrets(text: &str) -> String {
    let mut output = String::with_capacity(text.len());
    let mut token = String::new();

    for ch in text.chars() {
        if ch.is_whitespace() {
            if !token.is_empty() {
                output.push_str(&redact_secret_token(&token));
                token.clear();
            }
            output.push(ch);
        } else {
            token.push(ch);
        }
    }

    if !token.is_empty() {
        output.push_str(&redact_secret_token(&token));
    }

    output
}

fn redact_secret_token(token: &str) -> String {
    let lower = token.to_lowercase();
    if lower.starts_with("sk-")
        || lower.starts_with("ghp_")
        || lower.starts_with("xoxb-")
        || lower.starts_with("xoxa-")
        || lower.starts_with("xoxp-")
        || lower.starts_with("xoxr-")
        || lower.starts_with("xoxs-")
    {
        return "<redacted-secret>".to_string();
    }

    for marker in [
        "api_key", "api-key", "apikey", "secret", "token", "password",
    ] {
        if lower.starts_with(marker) && token.len() >= marker.len() + 17 {
            if let Some(separator) = token.find(['=', ':']) {
                let key = &token[..separator];
                return format!("{key}=<redacted>");
            }
        }
    }

    token.to_string()
}
